File: resource.go
package groundcover

import (
	"os"
	"runtime"
	"runtime/debug"
	"strconv"
	"time"
)

// OTel resource attribute keys used across resource detection and wire encoding.
const (
	attrServiceName  = "service.name"
	attrServiceVer   = "service.version"
	attrDeployEnv    = "deployment.environment.name"
	attrK8sNamespace = "k8s.namespace.name"
	attrK8sCluster   = "k8s.cluster.name"
)

// resource holds the detected resource/spine attributes, read once at init.
type resource struct {
	serviceName string
	env         string
	release     string
	namespace   string
	cluster     string

	// attrs are the OTel resource attributes carried with each event.
	attrs map[string]string
	// mainModule is the main module path used to classify in-app frames.
	mainModule string
	// startTime is the SDK initialization time, reported as session_start_time.
	startTime time.Time
}

// firstNonEmpty returns the first non-empty argument.
func firstNonEmpty(vals ...string) string {
	for _, v := range vals {
		if v != "" {
			return v
		}
	}
	return ""
}

// detectResource reads configuration, environment, and runtime information into
// a resource. Explicit Config fields take precedence over environment values.
func detectResource(cfg Config) resource {
	serviceName := firstNonEmpty(cfg.ServiceName, os.Getenv("OTEL_SERVICE_NAME"), os.Getenv("GC_SERVICE_NAME"))
	env := firstNonEmpty(cfg.Env, os.Getenv("GC_ENV"), os.Getenv("DEPLOYMENT_ENVIRONMENT"))
	release := firstNonEmpty(cfg.Release, os.Getenv("GC_RELEASE"))

	namespace := firstNonEmpty(os.Getenv("POD_NAMESPACE"), os.Getenv("NAMESPACE"), os.Getenv("K8S_NAMESPACE"))
	cluster := firstNonEmpty(os.Getenv("GC_CLUSTER"), os.Getenv("CLUSTER_NAME"), os.Getenv("K8S_CLUSTER_NAME"))
	podUID := os.Getenv("POD_UID")
	podName := firstNonEmpty(os.Getenv("POD_NAME"), os.Getenv("HOSTNAME"))
	nodeName := os.Getenv("NODE_NAME")

	hostname, _ := os.Hostname()

	mainModule := ""
	if bi, ok := debug.ReadBuildInfo(); ok && bi.Main.Path != "" {
		mainModule = bi.Main.Path
	}

	attrs := map[string]string{
		"telemetry.sdk.name":         sdkName,
		"telemetry.sdk.language":     sdkLanguage,
		"telemetry.sdk.version":      Version(),
		"process.runtime.name":       "go",
		"process.runtime.version":    runtime.Version(),
		"os.type":                    runtime.GOOS,
		"host.arch":                  runtime.GOARCH,
		"host.num_cpu":               strconv.Itoa(runtime.NumCPU()),
		"process.runtime.gomaxprocs": strconv.Itoa(runtime.GOMAXPROCS(0)),
	}
	putIfSet := func(k, v string) {
		if v != "" {
			attrs[k] = v
		}
	}
	putIfSet(attrServiceName, serviceName)
	putIfSet(attrServiceVer, release)
	putIfSet(attrDeployEnv, env)
	putIfSet("host.name", hostname)
	putIfSet("k8s.pod.uid", podUID)
	putIfSet("k8s.pod.name", podName)
	putIfSet(attrK8sNamespace, namespace)
	putIfSet("k8s.node.name", nodeName)
	putIfSet(attrK8sCluster, cluster)

	return resource{
		serviceName: serviceName,
		env:         env,
		release:     release,
		namespace:   namespace,
		cluster:     cluster,
		attrs:       attrs,
		mainModule:  mainModule,
		startTime:   time.Now(),
	}
}

File: groundcover-python/src/groundcover/_resource.py
"""Resource detection: service identity and runtime/host/k8s attributes."""

from __future__ import annotations

import dataclasses
import os
import platform
import socket
import sys
import time
from typing import Dict

from ._config import Config
from ._version import SDK_LANGUAGE, SDK_NAME, VERSION

# OTel resource attribute keys used across resource detection and wire encoding.
ATTR_SERVICE_NAME = "service.name"
ATTR_SERVICE_VER = "service.version"
ATTR_DEPLOY_ENV = "deployment.environment.name"
ATTR_K8S_NAMESPACE = "k8s.namespace.name"
ATTR_K8S_CLUSTER = "k8s.cluster.name"


@dataclasses.dataclass
class Resource:
    """The detected resource/spine attributes, read once at init."""

    service_name: str = ""
    env: str = ""
    release: str = ""
    namespace: str = ""
    cluster: str = ""

    attrs: Dict[str, str] = dataclasses.field(default_factory=dict)
    """The OTel resource attributes carried with each event."""
    in_app_root: str = ""
    """The application root path used to classify in-app frames."""
    start_time_ns: int = 0
    """The SDK initialization time, reported as session_start_time."""


def first_non_empty(*vals: str) -> str:
    """Return the first non-empty argument."""
    for v in vals:
        if v:
            return v
    return ""


def _detect_app_root() -> str:
    """Best-effort detection of the application root: the directory of the
    ``__main__`` module's file, falling back to the working directory. This is
    the Python analog of the Go SDK reading the main module path from build
    info."""
    main = sys.modules.get("__main__")
    main_file = getattr(main, "__file__", None)
    if main_file:
        try:
            return os.path.dirname(os.path.abspath(main_file))
        except Exception:
            pass
    try:
        return os.getcwd()
    except Exception:
        return ""


def detect_resource(cfg: Config) -> Resource:
    """Read configuration, environment, and runtime information into a
    Resource. Explicit Config fields take precedence over environment
    values."""
    env_get = os.environ.get

    service_name = first_non_empty(
        cfg.service_name, env_get("OTEL_SERVICE_NAME", ""), env_get("GC_SERVICE_NAME", "")
    )
    env = first_non_empty(cfg.env, env_get("GC_ENV", ""), env_get("DEPLOYMENT_ENVIRONMENT", ""))
    release = first_non_empty(cfg.release, env_get("GC_RELEASE", ""))

    namespace = first_non_empty(
        env_get("POD_NAMESPACE", ""), env_get("NAMESPACE", ""), env_get("K8S_NAMESPACE", "")
    )
    cluster = first_non_empty(
        env_get("GC_CLUSTER", ""), env_get("CLUSTER_NAME", ""), env_get("K8S_CLUSTER_NAME", "")
    )
    pod_uid = env_get("POD_UID", "")
    pod_name = first_non_empty(env_get("POD_NAME", ""), env_get("HOSTNAME", ""))
    node_name = env_get("NODE_NAME", "")

    try:
        hostname = socket.gethostname()
    except Exception:
        hostname = ""

    in_app_root = cfg.in_app_root or _detect_app_root()

    attrs: Dict[str, str] = {
        "telemetry.sdk.name": SDK_NAME,
        "telemetry.sdk.language": SDK_LANGUAGE,
        "telemetry.sdk.version": VERSION,
        "process.runtime.name": platform.python_implementation().lower(),
        "process.runtime.version": platform.python_version(),
        "os.type": sys.platform,
        "host.arch": platform.machine(),
        "host.num_cpu": str(os.cpu_count() or 0),
    }

    def put_if_set(k: str, v: str) -> None:
        if v:
            attrs[k] = v

    put_if_set(ATTR_SERVICE_NAME, service_name)
    put_if_set(ATTR_SERVICE_VER, release)
    put_if_set(ATTR_DEPLOY_ENV, env)
    put_if_set("host.name", hostname)
    put_if_set("k8s.pod.uid", pod_uid)
    put_if_set("k8s.pod.name", pod_name)
    put_if_set(ATTR_K8S_NAMESPACE, namespace)
    put_if_set("k8s.node.name", node_name)
    put_if_set(ATTR_K8S_CLUSTER, cluster)

    return Resource(
        service_name=service_name,
        env=env,
        release=release,
        namespace=namespace,
        cluster=cluster,
        attrs=attrs,
        in_app_root=in_app_root,
        start_time_ns=time.time_ns(),
    )
